### PR TITLE
﻿fix(rate-limit): add fast-path pre-check to avoid write amplificatio…

### PR DESCRIPTION
**Pull request by itsfaryal**
## Summary

Refs #367

Sustained over-limit traffic (credential stuffing, signup spam) currently costs 2
writes per rejected request via the existing optimistic-increment-then-compensate
path. This adds a zero-write fast-path for requests already clearly over the limit,
without touching the existing concurrency guarantee at the boundary.

## What changed

- Added a fast-path pre-check using the raw `windowCount` field, gated on
  `windowStart === currentWindowStart` to avoid false rejections from a stale
  record in an expired window.
- If `windowCount >= rule.limit` for the current window, reject immediately with
  zero writes.
- The existing optimistic-increment-then-compensate logic is unchanged for
  everything that doesn't clear the fast-path threshold (i.e. near the boundary,
  where previous-window weighting matters).

## Why this is safe

Previous-window weight is non-negative, so if the raw `windowCount` alone already
meets/exceeds the limit, the true weighted count does too — no ambiguity, no
atomicity required for this case. This targets only requests that are already
unambiguously over limit, not the boundary case #367 was written to fix.

## Testing

- Added test: fast-path rejects a request with zero writes when over limit
- Added test: a request within the fast-path threshold still resolves via the
  existing weighted-window logic (unchanged behavior)
- `npx vitest run src/rate-limit/rules.test.ts` — all passing
- `npm run lint:check` — clean
- `npm run build` — clean

---

> [!NOTE]
> **Medium Risk**
> Changes core rate-limit persistence and rejection ordering; incorrect fast-path gating could mis-reject or amplify writes, though boundary behavior is explicitly tested.
> 
> **Overview**
> Reduces database write amplification when traffic is already clearly over the limit (e.g. login/signup abuse).
> 
> **`consumeRateLimit`** now rejects immediately after `findOne` when the record is in the **current** window and raw `windowCount >= limit`—**no** `upsertOne`/`updateOne`. Near-boundary cases still use **optimistic increment** (`upsertOne` first), then reject with a **compensating decrement** (`updateOne` with `windowCount: { $gt: 0 }`) when the weighted count would exceed the limit (`count + 1 > rule.limit`).
> 
> Tests mock `updateOne` and cover sustained rejected bursts (zero writes), sliding-window boundary (increment + decrement), and concurrent rollover (counts never below zero).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a1ee72a4ac145b7e012c3be9d04e4a49cdf8d63f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/modelence/src/rate-limit/rules.ts [hunk 1]
```diff
@@ -76,15 +85,12 @@ async function checkRateLimitRule(rule: RateLimitRule, value: string, createErro
         },
       };
 
-  if (count >= rule.limit) {
+  await dbRateLimits.upsertOne(filter, modifier);
+
+  if (count + 1 > rule.limit) {
+    await dbRateLimits.upsertOne(filter, { $inc: { windowCount: -1 } });
```
[the remark this answers is not in the document]
Comment (itsfaryal) [reply packages/modelence/src/rate-limit/rules.ts]: Fixed! decrement now uses updateOne with a windowCount: { $gt: 0 } filter guard, so it silently no-ops once count hits 0 instead of going negative. See commit a1ee72a.
